Make API fields immutable

matte21 opened the pull request:
Make Subnets and Network attachments spec fields immutable.
Revamp KOS controllers to leverage immutability. 

Also, perform some fixes to code and comments unrelated to immutability.

Prominent or tricky changes have a comment to give more information.

matte21 committed:
Make Subnet and NetworkAttachment specs immutable

Change certain Subnet and NetworkAttachment spec fields by making them
immutable. This is conceptually sound as such fields (e.g. subnet CIDR
block) are the epitome of Subnets and NetworkAttachments.
The change allows slightly simpler controller code, but some headaches
associated with mutability are unavoidable because an API object could
be quickly deleted and recreated with the same namespaced name and
different spec fields' values.

matte21 committed:
Make subnet validator use subnet spec immutability

Make reference enqueuing in subnet update notification handlers
conditional on a UID change. Subnet updates are irrelevant to the
subnet validator: VNI and CIDR, the only information it cares about,
are immutable. The only exception are update notifications triggered
by a delete followed by a re-create. But these entail a UID change.

Also, make additions to conflicts caches conditional to equality
between in-cache owner's UID and live owner UID's. Previously a
check on in-cache owner's VNI and CIDR and live owner's VNI and CIDR
were used. Using UIDs is more straightforward and is made possible
by the fact that VNI and CIDR are allowed to change only when there's
a delete followed by a re-create, which entails a UID change.

matte21 committed:
Update IPAM after API fields immutability

Replace subnet resource version with subnet UID in anticipation data
used to remember in-flight updates and avoid duplicate work. The
anticipation data should be reset if a subnet's CIDR or VNI changes,
and that's why the resource version was needed. After immutability,
CIDR and VNI can only be changed by deleting and re-creating, and this
entails a UID change. This is more efficient because resource version
also changes on updates which do not change CIDR and VNI.

matte21 committed:
Update conn-agent after API fields immutability

Remove from the field selector for remote attachments the constraint
that status.hostIP is different than the local hostIP. It was needed
because updates to a local attachment X's spec.node made X transition
from local to remote, while keeping status.hostIP set to the local
hostIP. This could lead to keeping a local interface for a remote
attachment. With immutability, this is no longer needed because if X
is deleted and re-created in a way which makes it transition from local
to remote its status gets reset.

Also, change update notification handlers to enqueue references only if
the update is relevant.

Also, fix a bug in initial interfaces sync that could cause leaks of
interfaces.

Also, revamp some comments and do some minor refactorings.

matte21 committed:
Get rid of merge conflicts from IPAM

matte21 commented on staging/kos/pkg/apis/network/types.go:
```diff
@@ -170,42 +173,37 @@ type NetworkAttachment struct {
-	Errors SubnetErrors
-}
-
-type SubnetErrors struct {
```
Unrelated to immutability: we had a SubnetErrors type to wrap and distinguish errors from IPAM and subnet validator. We agreed that having the IPAM write errors in subnet status is not worth it. Thus the SubnetErrors type is removed and replaced by a slice of strings renamed Errors: there's no ambiguity on who populates it, it's always the validator. We can give it a more descriptive name if you want, such as `validatorErrors`, but I think it is ok to leave things unchanged.

matte21 commented on staging/kos/pkg/controllers/connectionagent/connectionagent.go:
```diff
@@ -432,13 +432,14 @@ func (ca *ConnectionAgent) onLocalAttAdded(obj interface{}) {
+func (ca *ConnectionAgent) onLocalAttUpdated(oldObj, obj interface{}) {
+	oldAtt, att := oldObj.(*netv1a1.NetworkAttachment), obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of update from %#+v to %#+v", oldAtt, att)
+	if oldAtt.Status.IPv4 != att.Status.IPv4 || oldAtt.Status.AddressVNI != att.Status.AddressVNI {
```
For efficiency enqueue only if the update can lead to a change of the interface. This was possible even without immutability, but the if condition would have been longer.

matte21 commented on staging/kos/pkg/controllers/connectionagent/connectionagent.go:
```diff
@@ -1126,14 +1126,14 @@ func (ca *ConnectionAgent) onRemoteAttAdded(obj interface{}) {
+func (ca *ConnectionAgent) onRemoteAttUpdated(oldObj, obj interface{}) {
+	oldAtt, att := oldObj.(*netv1a1.NetworkAttachment), obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Remote NetworkAttachments cache for VNI %06x: notified of update from %#+v to %#+v.", att.Status.AddressVNI, oldAtt, att)
+	if oldAtt.Status.IPv4 != att.Status.IPv4 || oldAtt.Status.HostIP != att.Status.HostIP {
```
For efficiency enqueue only if the update can lead to a change of the interface. This was possible even without immutability, but the if condition would have been longer.

matte21 commented on staging/kos/pkg/controllers/connectionagent/connectionagent.go:
```diff
@@ -575,21 +576,21 @@ func (ca *ConnectionAgent) syncPreExistingRemoteIfcs() error {
-					// Do not run the PostDeleteExec, it probably
-					// would not work because the PostCreateExec ran
-					// in a different container if at all.
+					ca.unsetLocalAttState(nsn)
```
Unrelated to immutability: this instruction applies to an attachment that transitions from local to remote during the initial interfaces sync from the connection agent. It unsets from an attachment's state its old local interface which has been deleted and replaced by a new remote interface during the initial sync. There was a comment from you asking why it was missing. You had gotten it right, its absence caused a bug. If that instruction is missing and the attachment is updated and a new interface is needed, the worker processing the attachment finds the old local interface in the attachment state and deletes it. The interface has already been deleted and the operation fails silently. The problem is that deletion of old/stale interfaces was mutually exclusive between local and remote interfaces: if a stale local interface was found it was assumed that there could not be also a stale remote interface for the same attachment. This mixed with the absence of the function this comment is about caused interfaces leaks.

MikeSpreitzer replied:
Yes.
Reviewing this reminded me of how complicated this logic is.  I am re-motivated to look at simplifying it, in a later PR.

MikeSpreitzer replied:
I also noticed that all the commentary about the bool returned from `updateVNStateForExistingAtt` is wrong, it speaks of `false` where `true` is meant.

matte21 replied:
> I also noticed that all the commentary about the bool returned from updateVNStateForExistingAtt is wrong, it speaks of false where true is meant.

Dang, thanks for pointing it out. Fixed in 7690042.

matte21 replied:
> Yes.
> Reviewing this reminded me of how complicated this logic is. I am re-motivated to look at simplifying it, in a later PR.

Are you referring to the whole CA or the initial sync up only? I agree that the whole CA is complicated, I think the sync up logic is not that difficult if properly explained (maybe better comments could be added) but is not apparent from that code. As I said while I was doing this and other PRs I stashed away unrelated tweaks to all controllers, I was thinking about merging them in a PR after this one. A considerable chunk of the CA tweaks is about the initial sync up (but they are not simplifications). That PR review might be the place to discuss changes. I might be wrong about this, but I fear there's not much complexity we can get rid of in that code.

MikeSpreitzer replied:
I think the whole CA is more complicated than it needs to be, including the initial sync.

matte21 commented on staging/kos/pkg/controllers/connectionagent/connectionagent.go:
```diff
@@ -970,7 +970,8 @@ func (ca *ConnectionAgent) createOrUpdateIfc(attGuestIP, attHostIP gonet.IP, att
 			}
 			ca.LaunchCommand(attNSN, &oldLocalState.LocalNetIfc, oldLocalState.PostDeleteExec, "postDelete", true, false)
 			ca.unsetLocalAttState(attNSN)
-		} else if attHasRemoteIfc {
```
Unrelated to immutability: this change makes the code look for and delete stale local and remote interfaces independently as opposed to in mutual exclusion. An attachment should never own more than one interface, but in case there's a bug somewhere in the code this change makes the connection agent self-heal. This change was introduced as a consequence of the problem described in the previous comment.

MikeSpreitzer replied:
good

matte21 commented on staging/kos/pkg/controllers/connectionagent/connectionagent.go:
```diff
@@ -1302,16 +1302,6 @@ func (ca *ConnectionAgent) remoteAttInVNWithVirtualIPHostIPSelector(vni uint32)
-	// connection node (status.hostIP is set with an update by the connection
-	// agent on the node of the attachment). This could result in the creation
-	// of a remote interface with the host IP of the connection agent (local) node.
-	hostIPIsNotLocal := k8sfields.OneTermNotEqualSelector(attHostIP, ca.hostIP.String())
```
This selector was needed because if spec.Node was updated causing an attachment to transition from local to remote, the fact that status.hostIP still stored the local hostIP caused transients during which a remote attachment had a remote interface with a local host IP.  With immutability this is not needed, an attachment’s node can change only through a delete and a re-create, which will reset the status, including hostIP.

matte21 commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -109,24 +109,22 @@ type IPAMController struct {
 	anticipatingResourceVersion string
 	anticipatedResourceVersion  string
-	anticipationSubnetRV        string
+	anticipationSubnetUID       k8stypes.UID
```
anticipation data to remember in-flight status updates was reset if the recorded subnet RV did not match the last seen subnet RV. Such data should be reset if a subnet VNI or CIDR change. After immutability such change can happen only if the subnet UID changes => replace subnet RV with subnet UID in anticipation data: there are less useless resets of the anticipation data.

matte21 commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -46,36 +47,32 @@ import (
-	// a conflict with X was found and at that time X had the VNI and CIDR
-	// values in ownerSummary.
+	// ownerUID is the UID of the subnet owning the conflicts cache.
+	ownerUID k8stypes.UID
```
Adding a subnet X to another subnet Y’s conflicts cache was conditional on Y’s VNI and CIDR recorded in the conflicts cache matching those in the Y received from the API server. Since after immutability Y’s VNI and CIDR can change only if Y’s UID changes, compare UIDs instead of VNI and CIDR --- it’s less fine-grained but simpler.

matte21 commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -156,8 +152,11 @@ func (v *Validator) OnUpdate(oldObj, newObj interface{}) {
+	// and CIDR block affect validation and they're immutable. But a subnet
+	// could still be deleted and re-created with different VNI and/or CIDR
+	// block, and the two events could be delivered as an update notification.
+	if oldS.UID != newS.UID {
```
The fields that matter towards validation can only change if UID changes: enqueue only if that's the case.

matte21 commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -228,33 +227,33 @@ func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
 
 	if v.subnetIsStale(ss.NamespacedName, s.ResourceVersion) {
-		return nil
+		return fmt.Errorf("current version (%s) is stale", s.ResourceVersion)
```
Returning nil was a bug because it assumed a new reference to the attachment would be enqueued upon receiving the fresh version of the subnet. But the fresh version of the subnet is delivered through an update notification which does not perform an enqueue, because enqueues are conditional to a change in UID (or  VNI and CIDR before this PR).

MikeSpreitzer replied:
(A) requeuing in this case is effectively polling for a notification.  Normally the notification should come quickly --- perhaps even more quickly than the retry; in bad weather the notification might not arrive for a while.  While not incorrect, I suspect polling for notification is not the best approach.

(B) Since a SubnetSpec is entirely immutable, the Subnet can need new processing only if the UID has changed, right?

matte21 replied:
I'll give a unique answer for (A) and (B).

A subnet X needs re-processing also if a rival has been deleted.
This is achieved by recording X in a rival Y's conflicts cache. Upon Y's deletion X is enqueued and reprocessed. But if X was in Y's conflicts cache there's a conflict and its `status.errors` field has been updated to reflect that. If Y is deleted and X is re-processed before X's updated version appears in the Informer's cache, re-processing of X will fail because the new status update will fail, the cause being a resource version mismatch. This is what happens if we remove entirely the check. It is not incorrect: the error from the server will cause a delayed re-processing of the subnet and hopefully by then the Informer's cache records the up to date version of X.

Having that check (and returning an error if the subnet is stale) achieves the exact same result without the call to the API server, which is good for performance (the idea I got by reading PRs and issues in K8s and docs is that K8s people are rather protective with the API servers --- if a call can be easily avoided it should be avoided). OTOH, conflicts should be extremely rare and are always caused by a user's mistake.

You're right that retrying is polling for a notification. Not retrying though is incorrect because the notification will not enqueue anything because UID did not change: a needed processing is never performed. A possible path out of this could be changing the condition in the update notification handler which decides when there is an enqueue, but I don't like the idea and that is still polling for a notification.

MikeSpreitzer replied:
Right.

The suppression of work on stale objects is correct, because any work on a stale object is guaranteed to fail.

In the handler for an update notification, we have been focused on the question of whether the update indicates a change in object content that matters.  But an update also indicates a change in ResourceVersion, and that alone can matter.  A correct test in the update notification handler would be whether either (a) there is a change in UID or (b) the object is in the stale cache with a ResourceVersion different from the new ResoureVersion.

That updated test is more liberal than we need; we could narrow it to test whether either (a) there is a change in UID or (b) both (b.1) the object is in the stale cache with a ResourceVersion different from the new ResoureVersion and (b.2) the object needs to be checked because it was recently evicted from a conflict cache.   But I think the added complexity is not worth it.  The more liberal test is not incorrect, and the restriction would save noticeable work only in the case of conflicts, which should be rare, so the restriction would have little performance benefit.

matte21 replied:
The problem with your solution (the liberal one) is that a conflicting subnet X keeps being validated indefinitely. Its status is updated to record the conflicts, the stale RV is written in the cache. When the update notification is received the subnet namespaced name is enqueued and the subnet is processed again, and this keeps happening indefinitely because there will be a new status update. A way out is the conservative solution you outlined, but as you said it is complex (much more than the current solution --- returning an error and retrying). Also, such solution is not conceptually better than polling for a notification: you're waiting for a specific notification. You're stating that specific events happened and you're waiting for a notification which lets you process them. Another way out is making the status update conditional to whether the update would actually change the status, and if that's not the case returning nil and stopping processing. This way an already updated conflicting subnet is processed only once more than necessary. But the status update is the last step of processing: all the work has already been done (including the expensive list to the API server). I'd rather have work suppression suppress work ASAP.

matte21 replied:
I've been thinking about how to implement the conservative solution. Currently the stale RVs cache is a map from namespaced name to stale RV. Replace the map value with a struct storing the stale RV and a bool saying whether the namespaced name has been evicted from a conflicts cache. The bool is set to true when the subnet is evicted from a conflicts cache. The update notification handler enqueues if the UIDs differ or if the subnet is not stale and the eviction flag is true.

I'm not against leaving things the way they are (returning an error in case of stale subnet AKA polling for a notification). But if you are convinced that we should do something different, I don't think the more conservative solution you outlined adds that much complexity.

MikeSpreitzer replied:
You touched on another solution that is simple and principled: request an update only if there is actually a change being requested.  Yes, that lets in one round of unnecessary work.  But that should be rare (because conflicts are rare), so I am inclined to stop there.  Leave the update notification handler to simply test whether (a) the UID has changed or (b) there is a doomed entry in the stale RV cache.

matte21 replied:
Done in dc54563.

matte21 commented on staging/kos/pkg/util/parse/network/subnet/summary.go:
```diff
@@ -37,6 +37,7 @@ const (
 // tasks.
 type Summary struct {
 	NamespacedName types.NamespacedName
+	UID            types.UID
```
Subnet summary is simplified by immutability: it needs to store the UID, but some functions become useless and can be removed.

matte21 commented on staging/kos/pkg/registry/network/subnet/strategy.go:
```diff
@@ -167,15 +137,13 @@ func (ss *subnetStrategy) checkNSAndCIDRConflicts(candidate *subnet.Summary, mal
-		glog.V(6).Infof("Validating %s against %s", candidate.NamespacedName, pr.NamespacedName)
+		// Ignore the error because a malformed subnet would not have been
+		// allowed: it's the code in this file that performs validation.
+		pr, _ := subnet.NewSummary(potentialRival)
```
A PR with additional tweaks will follow: the error will be checked. Ignoring it does not account for cases where this version of the API server is rolled out after a more lenient version which lets malformed Subnets through.

MikeSpreitzer replied:
I would not worry about that, we are not in production.

matte21 commented on staging/kos/pkg/registry/network/subnet/strategy.go:
```diff
@@ -167,15 +137,13 @@ func (ss *subnetStrategy) checkNSAndCIDRConflicts(candidate *subnet.Summary, mal
+	// Check whether there are Namespace and CIDR conflicts with other subnets.
 	for _, potentialRival := range potentialRivals {
-		pr, err := subnet.NewSummary(potentialRival)
-		if err != nil || candidate.SameSubnetAs(pr) {
```
No need to check whether `potentialRival` is the same subnet as the one under validation and skip it if that's the case because after immutability this code is only executed upon subnet creation.

MikeSpreitzer commented on staging/kos/pkg/apis/network/types.go:
```diff
@@ -170,42 +173,37 @@ type NetworkAttachment struct {
-	// Validation holds the complaints, if any, from the subnets validator. It
-	// might contain an explanation on why SubnetStatus.Validated is false.
+	// Errors holds the complaints, if any, from the subnets validator. It might
+	// contain an explanation on why SubnetStatus.Validated is false.
```
I think this comment should make a stronger assertion.  It should say that this slice is non-empty if and only if `Validated==false` for reasons (2) or (3).

matte21 replied:
Agreed, done in 7690042.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -228,33 +227,33 @@ func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
+	// case the conflicts cache still records a deleted subnet with same
+	// namespaced name as s return old rivals and reconsider them: the conflict
+	// might have disappeared.
+	oldRivals := v.initConflictsCache(ss.NamespacedName, ss.UID)
```
The lead sentence of the comment, and the function name, have been changed to refer only to the init case.  But that is not the only case.

matte21 replied:
Rolled back function name and tweaked comment in 7690042.
The rationale for the name change was that since updates to a subnet are forbidden when that function is executed and a new conflicts cache must be added it is always being initialized for a new subnet. When we had updates, a new conflicts cache could be added (after deleting the stale one)  because of an update to the same subnet. But practically speaking there's no difference: we still have the removal of an old conflicts cache and the addition of a new one, which is equivalent to an update, so I reverted the change.

MikeSpreitzer commented on staging/kos/pkg/registry/network/networkattachment/strategy.go:
```diff
@@ -107,5 +109,29 @@ func (networkattachmentStrategy) Canonicalize(obj runtime.Object) {
+	immutableFieldMsg := "attempt to update immutable field"
+	newNa, oldNa := obj.(*network.NetworkAttachment), old.(*network.NetworkAttachment)
+	if newNa.Spec.Node != oldNa.Spec.Node {
+		errs = append(errs, field.Forbidden(field.NewPath("spec.node"), immutableFieldMsg))
```
It should be `field.NewPath("spec", "node")`

matte21 replied:
Dang, fixed in 7690042.

matte21 committed:
Fix problems emerged in PR review

matte21 committed:
Change stale subnets handling in subnet validator

Make subnet validator suppress useless work due to stale subnets w/o
relying on notification polling. This entails tweaking the update
notification handler to enqueue in case the notification carries a
fresh subnet which was previously stale and making subnets' status
updates conditional on whether the status has changed.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -156,13 +155,30 @@ func (v *Validator) OnUpdate(oldObj, newObj interface{}) {
+	}
+}
+
+func (v *Validator) subnetIsAwaited(subnet k8stypes.NamespacedName, rv string) bool {
```
The name "subnetIsAwaited" has positive connotations, whereas the actual condition here is more of a negative.  I suggest a different name, perhaps something like "subnetStaleNoMore", and/or adding an explanatory comment.

MikeSpreitzer replied:
This can be handled in a follow-up PR.
